shcomplete/seq2seq_correction.py

```diff
@@ -0,0 +1,225 @@
+
+
+def generate_model(chars, gConfig):
+    """
+    Generate the model.
+    """
+    model = Sequential()
+    for layer_number in range(gConfig["input_layers"]):
+        model.add(recurrent.LSTM(gConfig["hidden_layers"], input_shape=(None, len(chars)),
+                                 return_sequences=layer_number + 1 < gConfig["input_layers"]))
+        model.add(Dropout(gConfig["amount_of_dropout"]))
+
+    model.add(RepeatVector(gConfig["max_cmd_len"]))
+    for _ in range(gConfig["output_layers"]):
+        model.add(recurrent.LSTM(gConfig["hidden_layers"], return_sequences=True))
+        model.add(Dropout(gConfig["amount_of_dropout"]))
+
+    model.add(TimeDistributed(Dense(len(chars))))
+    model.add(Activation('relu'))
+    model.compile(loss='mse', optimizer='adam')
+    return model
+
+
+def generate_input(chars, command, vocab_trie, gConfig):
+    """
+    Generate a tuple (misspelled_command, true_command) to train the model.
+    The misspelled_command is generated by adding artificial mistakes.abs
+    Padding is added to feed the model with fixed length sequences of characters.
+    """
+    misspelled_command = create_mistakes(chars, command, vocab_trie, gConfig["level_of_noise"])
+    misspelled_command += gConfig["padding"] * (gConfig["max_cmd_len"] - len(misspelled_command))
+    command += gConfig["padding"] * (gConfig["max_cmd_len"] - len(command))
+    return misspelled_command, command
+
+
+def sample_prediction(model, chars, X, y, max_cmd_len, inverted):
+    """
+    Select 10 misspelled commands and print the current correction of the model.
+    """
+    seq2seq = Seq2seq(chars)
+    print()
+    for _ in range(10):
+        ind = randint(0, len(X))
+        rowX = X[np.array([ind])]
+        rowy = y[np.array([ind])]
+        preds = model.predict_classes(rowX)
+        true_command = seq2seq.decode(rowy[0], max_cmd_len)
+        model_correction = seq2seq.decode(preds[0], max_cmd_len, reduction=False)
+        if inverted:
+            misspelled_command = seq2seq.decode(rowX[0], max_cmd_len, inverted=inverted)
+            print('Command misspelled :', misspelled_command[::-1])
+        else:
+            print('Q :', misspelled_command)
+        print('True command :', true_command)
+        print("Correction predicted :", model_correction)
+        print('---')
+    return model_correction
+
+
+class Seq2seq(object):
+
+    def __init__(self, chars):
+        self.chars = sorted(set(chars))
+        self.char_indices = dict((c, i) for i, c in enumerate(self.chars))
+        self.indices_char = dict((i, c) for i, c in enumerate(self.chars))
+
+    def encode(self, commands, max_cmd_len):
+        """
+        Encode commands into numpy arrays.
+        """
+        X = np.zeros((len(commands), max_cmd_len, len(self.chars)), dtype=np.bool)
+        for i, cmd in enumerate(commands):
+            for j, char in enumerate(cmd):
+                try:
+                    X[i, j, self.char_indices[char]] = 1
```

Comment:
I have a feeling that Keras has this one hot encoder. https://machinelearningmastery.com/how-to-one-hot-encode-sequence-data-in-python/